src/scripts/generate_vaccination_status_datamart.py

```diff
@@ -0,0 +1,224 @@
+                and current_dose_age > 12
+            ):
+                vaccinated_date = check_mrna_status(
+                    idx,
+                    patient_immunizations,
+                    vax_codes["pfizer_adult"],
+                    vax_codes["pfizer_range"],
+                )
+                initial_vax_type = "pfizer"
+                next_dose_idx = idx + 2
+
+            # Pfizer ages 5 to <12
+            if vaccination.vaccineCode in vax_codes[
+                "pfizer_pediatric"
+            ] and current_dose_age in range(5, 12):
+                vaccinated_date = check_mrna_status(
+                    idx,
+                    patient_immunizations,
+                    vax_codes["pfizer_pediatric"],
+                    vax_codes["pfizer_range"],
+                )
+                initial_vax_type = "pfizer"
+                next_dose_idx = idx + 2
+
+        # Check for booster
+        if pd.notnull(vaccinated_date) and len(patient_immunizations) > next_dose_idx:
+            current_dose_age = (
+                patient_immunizations.occurrenceDateTime[next_dose_idx]
+                - vaccination.birthDate
+            ).days / 365.25
+            if (
+                current_dose_age >= 18
+                and patient_immunizations.vaccineCode[next_dose_idx]
+                in vax_codes["adult_boosters"]
+                and (
+                    patient_immunizations.occurrenceDateTime[next_dose_idx]
+                    - vaccination.occurrenceDateTime
+                ).days
+                > vax_codes[initial_vax_type + "_booster_delay"]
+            ):
+                boosted_date = patient_immunizations.occurrenceDateTime[next_dose_idx]
+                break
+            elif (
+                current_dose_age in range(12, 18)
+                and patient_immunizations.vaccineCode[next_dose_idx]
+                in vax_codes["pfizer_12to17_boosters"]
+                and (
+                    patient_immunizations.occurrenceDateTime[next_dose_idx]
+                    - vaccination.occurrenceDataTime
+                ).days
+                > vax_codes[initial_vax_type + "_booster_delay"]
+            ):
+                boosted_date = patient_immunizations.occurrenceDateTime[next_dose_idx]
+                break
+
+    patient_with_vax_status = patient_immunizations.agg(record_combination_func)
+    patient_with_vax_status["vaccinationDate"] = vaccinated_date
+    patient_with_vax_status["boosterDate"] = boosted_date
+    return patient_with_vax_status
+
+
+def check_mrna_status(
+    idx: int, immunizations: pd.DataFrame, codes: list, second_dose_range: range
+) -> datetime.datetime:
+    """
+    Given a patient's immunization records, a vax code list, and a second dose range,
+    determine the date when the patient completed an initial vaccination sequence for
+    COVID with an mRNA vaccine.
+    """
+    vaccinated_date = pd.NaT
+    if len(immunizations) > idx + 1:
```

Comment:
I wonder if some of the assumptions we are making are valid.  Here we are assuming the dose chronologically after the current dose is the 2nd... Could there be an edge case where a person went a week later the first dose by mistake/based on bad info, and that one didn't count, but they went again in the true second dose range.  I don't think this logic would accept that right?  It's just something in the back of my mind - I do think this is good enough for our purposes right now.

Reply:
This is exactly the sort of logic that would have to be driven, by the specific analysis question at hand. This approach applies the official definition of "fully vaccinated" very strictly. Obviously for real people in the real world life is a lot messier. Definitely something to discuss with someone who was interested in actually use some of this code in production.